### ISSUE TITLE
ExtUtils-ParseXS-3.44: xsubpp seems to output false-positive warnings

### ISSUE THREAD

**Issue by twata1** [Perl/perl5#19661]
Module:
ExtUtils-ParseXS-3.44

**Description**

Hello,

The following link is the background.

[DBD-MariaDB-1.22: Build warnings](https://github.com/perl5-dbi/DBD-MariaDB/issues/169)

As shown below, the **xsubpp** command outputs warnings when I run the **gmake** command.
According to https://github.com/perl5-dbi/DBD-MariaDB/issues/169#issuecomment-1107666139 here, it seems to be a false positive warnings.

```
Warning: duplicate function definition 'do' detected in MariaDB.xs, line 104
Warning: duplicate function definition 'rows' detected in MariaDB.xs, line 229
Warning: duplicate function definition 'last_insert_id' detected in MariaDB.xs, line 248
```

**Steps to Reproduce**

0. Start the mysqld in advance for testing.
```
C:\mysql\bin>mysqld --defaults-file="..\my.ini" --console
```
1. Install ExtUtils::ParseXS 3.44 beforehand. (The following command confirms ExtUtils::ParseXS 's version.)
```
C:\home\sunlight2\DBD-MariaDB-1.22>module-version ExtUtils::ParseXS

The version of ExtUtils::ParseXS in C:\home\sunlight2\perl5\lib\perl5 is 3.44
```
2. Download DBD-MariaDB-1.22.tar.gz from [metacpan ](https://metacpan.org/dist/DBD-MariaDB)and extract it and run ``make Makefile.PL`` and ``gmake``.
```Perl
C:\home\sunlight2\DBD-MariaDB-1.22>perl Makefile.PL


PLEASE NOTE:

For 'make test' to run properly, you must ensure that the
database user 'root' can connect to your MariaDB or MySQL server
and has the proper privileges that these tests require such
as 'drop table', 'create table', 'drop procedure', 'create procedure'
as well as others.

mysql> grant all privileges on test.* to 'root'@'localhost' identified by 's3kr1t';

You can also optionally set the user to run 'make test' with:

perl Makefile.PL --testuser=username

I will use the following settings for compiling and testing:

  cflags       (mysql_config) = -IC:\Strawberry\c\bin\..\include\mysql50716
  libs         (mysql_config) = -LC:\Strawberry\c\bin\..\lib -lmysql
  mysql_config (guessed     ) = mysql_config
  testdb       (default     ) = test
  testhost     (default     ) =
  testpassword (default     ) =
  testport     (default     ) =
  testsocket   (default     ) =
  testuser     (guessed     ) = root

To change these settings, see 'perl Makefile.PL --help' and
'perldoc DBD::MariaDB::INSTALL'.

Checking if libs and header files are available for compiling...
Checking if correct version of MariaDB or MySQL client is present...
Looks good.

Embedded server: not supported by client library

Client library deinitialize OpenSSL library functions: no

Checking if your kit is complete...
Looks good
WARNING: Older versions of ExtUtils::MakeMaker may errantly install README.pod as part of this distribution. It is recom
mended to avoid using this path in CPAN modules.
Using DBI 1.643 (for perl 5.030002 on MSWin32-x64-multi-thread) installed in C:/Strawberry/perl/vendor/lib/auto/DBI/
Generating a gmake-style Makefile
Writing Makefile for DBD::MariaDB
"\x{0159}" does not map to cp932 at C:\home\sunlight2\perl5\lib\perl5/ExtUtils/MakeMaker.pm line 1266.
"\x{0159}" does not map to cp932 at C:\home\sunlight2\perl5\lib\perl5/ExtUtils/MakeMaker.pm line 1266.
"\x{0159}" does not map to cp932 at C:\home\sunlight2\perl5\lib\perl5/ExtUtils/MakeMaker.pm line 1266.
Writing MYMETA.yml and MYMETA.json

C:\home\sunlight2\DBD-MariaDB-1.22>gmake
cp lib/DBD/MariaDB.pod blib\lib/DBD/MariaDB.pod
cp lib/DBD/MariaDB/INSTALL.pod blib\lib/DBD/MariaDB/INSTALL.pod
cp lib/DBD/MariaDB.pm blib\lib/DBD/MariaDB.pm
cp README.pod blib\lib/DBD/MariaDB/README.pod
Running Mkbootstrap for MariaDB ()
"C:\Strawberry\perl\bin\perl.exe" -MExtUtils::Command -e chmod -- 644 "MariaDB.bs"
"C:\Strawberry\perl\bin\perl.exe" -MExtUtils::Command::MM -e cp_nonempty -- MariaDB.bs blib\arch\auto\DBD\MariaDB\MariaDB.bs 644
"C:\Strawberry\perl\bin\perl.exe" -p -e "s/~DRIVER~/MariaDB/g" C:\Strawberry\perl\vendor\lib\auto\DBI\Driver.xst > MariaDB.xsi
"C:\Strawberry\perl\bin\perl.exe" "C:\home\sunlight2\perl5\lib\perl5\ExtUtils/xsubpp"  -typemap C:\Strawberry\perl\lib\ExtUtils\typemap  MariaDB.xs > MariaDB.xsc
Warning: duplicate function definition 'do' detected in MariaDB.xs, line 104
Warning: duplicate function definition 'rows' detected in MariaDB.xs, line 229
Warning: duplicate function definition 'last_insert_id' detected in MariaDB.xs, line 248
"C:\Strawberry\perl\bin\perl.exe" -MExtUtils::Command -e mv -- MariaDB.xsc MariaDB.c
gcc -c  -IC:\Strawberry\perl\vendor\lib\auto\DBI -IC:\Strawberry\c\bin\..\include\mysql50716 -DHAVE_DBI_1_634 -DHAVE_DBI_1_642 -DHAVE_GET_OPTION -s -O2 -DWIN32 -DWIN64 -DCONSERVATIVE -D__USE_MINGW_ANSI_STDIO -DPERL_TEXTMODE_SCRIPTS -DPERL_IMPLICIT_CONTEXT -DPERL_IMPLICIT_SYS -DUSE_PERLIO -fwrapv -fno-strict-aliasing -mms-bitfields -s -O2   -DVERSION=\"1.22\" -DXS_VERSION=\"1.22\"  "-IC:\Strawberry\perl\lib\CORE"   MariaDB.c
gcc -c  -IC:\Strawberry\perl\vendor\lib\auto\DBI -IC:\Strawberry\c\bin\..\include\mysql50716 -DHAVE_DBI_1_634 -DHAVE_DBI_1_642 -DHAVE_GET_OPTION -s -O2 -DWIN32 -DWIN64 -DCONSERVATIVE -D__USE_MINGW_ANSI_STDIO -DPERL_TEXTMODE_SCRIPTS -DPERL_IMPLICIT_CONTEXT -DPERL_IMPLICIT_SYS -DUSE_PERLIO -fwrapv -fno-strict-aliasing -mms-bitfields -s -O2   -DVERSION=\"1.22\" -DXS_VERSION=\"1.22\"  "-IC:\Strawberry\perl\lib\CORE"   dbdimp.c
gcc -c  -IC:\Strawberry\perl\vendor\lib\auto\DBI -IC:\Strawberry\c\bin\..\include\mysql50716 -DHAVE_DBI_1_634 -DHAVE_DBI_1_642 -DHAVE_GET_OPTION -s -O2 -DWIN32 -DWIN64 -DCONSERVATIVE -D__USE_MINGW_ANSI_STDIO -DPERL_TEXTMODE_SCRIPTS -DPERL_IMPLICIT_CONTEXT -DPERL_IMPLICIT_SYS -DUSE_PERLIO -fwrapv -fno-strict-aliasing -mms-bitfields -s -O2   -DVERSION=\"1.22\" -DXS_VERSION=\"1.22\"  "-IC:\Strawberry\perl\lib\CORE"   socket.c
"C:\Strawberry\perl\bin\perl.exe" -MExtUtils::Mksymlists \
     -e "Mksymlists('NAME'=>\"DBD::MariaDB\", 'DLBASE' => 'MariaDB', 'DL_FUNCS' => {  }, 'FUNCLIST' => [], 'IMPORTS' =>{  }, 'DL_VARS' => []);"
g++ MariaDB.def -o blib\arch\auto\DBD\MariaDB\MariaDB.xs.dll -mdll -s -L"c:\Strawberry\perl\lib\CORE" -L"c:\Strawberry\c\lib" MariaDB.o dbdimp.o socket.o   "C:\Strawberry\perl\lib\CORE\libperl530.a" "C:\Strawberry\c\bin\..\lib\libmysql.a" "c:\Strawberry\c\x86_64-w64-mingw32\lib\libmoldname.a" "c:\Strawberry\c\x86_64-w64-mingw32\lib\libkernel32.a" "c:\Strawberry\c\x86_64-w64-mingw32\lib\libuser32.a" "c:\Strawberry\c\x86_64-w64-mingw32\lib\libgdi32.a" "c:\Strawberry\c\x86_64-w64-mingw32\lib\libwinspool.a" "c:\Strawberry\c\x86_64-w64-mingw32\lib\libcomdlg32.a" "c:\Strawberry\c\x86_64-w64-mingw32\lib\libadvapi32.a" "c:\Strawberry\c\x86_64-w64-mingw32\lib\libshell32.a" "c:\Strawberry\c\x86_64-w64-mingw32\lib\libole32.a" "c:\Strawberry\c\x86_64-w64-mingw32\lib\liboleaut32.a" "c:\Strawberry\c\x86_64-w64-mingw32\lib\libnetapi32.a" "c:\Strawberry\c\x86_64-w64-mingw32\lib\libuuid.a" "c:\Strawberry\c\x86_64-w64-mingw32\lib\libws2_32.a" "c:\Strawberry\c\x86_64-w64-mingw32\lib\libmpr.a" "c:\Strawberry\c\x86_64-w64-mingw32\lib\libwinmm.a" "c:\Strawberry\c\x86_64-w64-mingw32\lib\libversion.a" "c:\Strawberry\c\x86_64-w64-mingw32\lib\libodbc32.a" "c:\Strawberry\c\x86_64-w64-mingw32\lib\libodbccp32.a" "c:\Strawberry\c\x86_64-w64-mingw32\lib\libcomctl32.a" -Wl,--enable-auto-image-base
"C:\Strawberry\perl\bin\perl.exe" -MExtUtils::Command -e chmod -- 755 blib\arch\auto\DBD\MariaDB\MariaDB.xs.dll

C:\home\sunlight2\DBD-MariaDB-1.22>
```

**Expected behavior**

These warnings (``Warning: duplicate function definition ...``) are not output.

**Perl configuration**

```
C:\home\sunlight2\DBD-MariaDB-1.22>perl -V
Summary of my perl5 (revision 5 version 30 subversion 2) configuration:

  Platform:
    osname=MSWin32
    osvers=10.0.18363.720
    archname=MSWin32-x64-multi-thread
    uname='Win32 strawberry-perl 5.30.2.1 #1 Tue Mar 17 03:21:32 2020 x64'
    config_args='undef'
    hint=recommended
    useposix=true
    d_sigaction=undef
    useithreads=define
    usemultiplicity=define
    use64bitint=define
    use64bitall=undef
    uselongdouble=undef
    usemymalloc=n
    default_inc_excludes_dot=define
    bincompat5005=undef
  Compiler:
    cc='gcc'
    ccflags =' -s -O2 -DWIN32 -DWIN64 -DCONSERVATIVE -D__USE_MINGW_ANSI_STDIO -DPERL_TEXTMODE_SCRIPTS -DPERL_IMPLICIT_CO
NTEXT -DPERL_IMPLICIT_SYS -DUSE_PERLIO -fwrapv -fno-strict-aliasing -mms-bitfields'
    optimize='-s -O2'
    cppflags='-DWIN32'
    ccversion=''
    gccversion='8.3.0'
    gccosandvers=''
    intsize=4
    longsize=4
    ptrsize=8
    doublesize=8
    byteorder=12345678
    doublekind=3
    d_longlong=define
    longlongsize=8
    d_longdbl=define
    longdblsize=16
    longdblkind=3
    ivtype='long long'
    ivsize=8
    nvtype='double'
    nvsize=8
    Off_t='long long'
    lseeksize=8
    alignbytes=8
    prototype=define
  Linker and Libraries:
    ld='g++'
    ldflags ='-s -L"c:\Strawberry\perl\lib\CORE" -L"c:\Strawberry\c\lib"'
    libpth=c:\Strawberry\c\lib c:\Strawberry\c\x86_64-w64-mingw32\lib c:\Strawberry\c\lib\gcc\x86_64-w64-mingw32\8.3.0
    libs= -lmoldname -lkernel32 -luser32 -lgdi32 -lwinspool -lcomdlg32 -ladvapi32 -lshell32 -lole32 -loleaut32 -lnetapi32 -luuid -lws2_32 -lmpr -lwinmm -lversion -lodbc32 -lodbccp32 -lcomctl32
    perllibs= -lmoldname -lkernel32 -luser32 -lgdi32 -lwinspool -lcomdlg32 -ladvapi32 -lshell32 -lole32 -loleaut32 -lnetapi32 -luuid -lws2_32 -lmpr -lwinmm -lversion -lodbc32 -lodbccp32 -lcomctl32
    libc=
    so=dll
    useshrplib=true
    libperl=libperl530.a
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_win32.xs
    dlext=xs.dll
    d_dlsymun=undef
    ccdlflags=' '
    cccdlflags=' '
    lddlflags='-mdll -s -L"c:\Strawberry\perl\lib\CORE" -L"c:\Strawberry\c\lib"'


Characteristics of this binary (from libperl):
  Compile-time options:
    HAS_TIMES
    HAVE_INTERP_INTERN
    MULTIPLICITY
    PERLIO_LAYERS
    PERL_COPY_ON_WRITE
    PERL_DONT_CREATE_GVSV
    PERL_IMPLICIT_CONTEXT
    PERL_IMPLICIT_SYS
    PERL_MALLOC_WRAP
    PERL_OP_PARENT
    PERL_PRESERVE_IVUV
    USE_64_BIT_INT
    USE_ITHREADS
    USE_LARGE_FILES
    USE_LOCALE
    USE_LOCALE_COLLATE
    USE_LOCALE_CTYPE
    USE_LOCALE_NUMERIC
    USE_LOCALE_TIME
    USE_PERLIO
    USE_PERL_ATOF
  Built under MSWin32
  Compiled at Mar 17 2020 03:27:53
  %ENV:
    PERL5LIB="C:\home\sunlight2\perl5\lib\perl5"
    PERL_LOCAL_LIB_ROOT="C:\home\sunlight2\perl5"
    PERL_MB_OPT="--install_base "C:\\home\\sunlight2\\perl5""
    PERL_MM_OPT="INSTALL_BASE=C:\\home\\sunlight2\\perl5"
  @INC:
    C:\home\sunlight2\perl5\lib\perl5/MSWin32-x64-multi-thread
    C:\home\sunlight2\perl5\lib\perl5
    C:/Strawberry/perl/site/lib
    C:/Strawberry/perl/vendor/lib
    C:/Strawberry/perl/lib

C:\home\sunlight2\DBD-MariaDB-1.22>
```

**Comment by jkeenan** [comment]
> Module: ExtUtils-ParseXS-3.44
> 
> **Description**
> 
> Hello,
> 
> The following link is the background.
> 
> [DBD-MariaDB-1.22: Build warnings](https://github.com/perl5-dbi/DBD-MariaDB/issues/169)
> 
> As shown below, the **xsubpp** command outputs warnings when I run the **gmake** command. According to [perl5-dbi/DBD-MariaDB#169 (comment)](https://github.com/perl5-dbi/DBD-MariaDB/issues/169#issuecomment-1107666139) here, it seems to be a false positive warnings.
> 

@twata1, this ticket has not had a response in two months.  But that's not surprising:  The only way we can reproduce this is to set up MariaDB, `mysqld` and run `gmake`.  That's really out of scope for the Perl core distribution's test suite.

Can you provide some other, simpler illustration of this problem?  (If not, I recommend closing this ticket.)

**Comment by pali** [comment]
@jkeenan You can reproduce it also without MariaDB/MySQL server `mysqld`. You just need client libraries and then run perl Makefile.PL and make. Windows machine is not needed too, you can reproduce it also on Linux.

**Comment by jkeenan** [comment]
> @jkeenan You can reproduce it also without MariaDB/MySQL server `mysqld`. You just need client libraries and then run perl Makefile.PL and make. Windows machine is not needed too, you can reproduce it also on Linux.

@pali, thanks for looking into this.  Nonetheless, if MariaDB client libraries are required to demonstrate the problem, then it seems those libraries' issue tracker is the appropriate place to report the problem.

**Comment by pali** [comment]
So where is the issue tracker for `xsubpp` if not here? Could you instruct @twata1 where to report this issue? As was already stated in described I instructed @twata1 to move issue from module/library issue tracker to perl as this is not issue in module/library/MariaDB.

**Comment by choroba** [comment]
The behaviour can be easily replicated without any external libraries.
```
MODULE = My PACKAGE = My

#ifdef MYDEF123

void
do(dbh)
   SV *dbh
CODE:
{
   int x;
   ++x;
}

#endif

void
do(dbh)
   SV *dbh
CODE:
{
   int x;
   ++x;
}
```
Running `xsubpp` on this files generates the warning.

**Comment by jkeenan** [comment]
On 7/1/22 17:28, E. Choroba wrote:
> The behaviour can be easily replicated without any external libraries.
> 
> |MODULE = My PACKAGE = My #ifndef MYDEF123 void do(dbh) SV *dbh CODE: { 
> int x; ++x; } #endif void do(dbh) SV *dbh CODE: { int x; ++x; } |
> 
> Running |xsubpp| on this files generates the warning.
> 

Thanks; now we have something to work with!

**Comment by jkeenan** [comment]
> The behaviour can be easily replicated without any external libraries.
> [snip]

I have created this branch in our repository:  [smoke-me/jkeenan/parsexs-gh19661-20220704](https://github.com/Perl/perl5/tree/smoke-me/jkeenan/parsexs-gh19661-20220704).  In this branch I have added the XS file you provided and then have added one unit test to the EUPXS test suite.  That test is currently TODO-ed out because it detects the duplicate warning but no fix has yet been provided.  From that branch I have created Draft p.r. https://github.com/Perl/perl5/pull/19920.

So, anyone who wants to work on this bug can pick up a framework for doing so in that branch.

**Comment by pali** [comment]
@jkeenan Thanks! Anyway, for fixing this issue, somebody with XS compiler knowledge is needed. So I think it would not be easy.

**Comment by jkeenan** [comment]
It appears that the problem is not as simple as (as I would have thought from @choroba's sample file) using a string like `dbh` twice within an XS file, once inside an `#ifden`, once outside.

See 746c9d53a122e51f4b2509a3e78bf49d43bf577e in my branch.  We get the spurious warning even if we only use a function name like `dbh` once.

**Comment by sisyphus** [comment]
I think that @choroba's nice little demo shows that the warning is sort-of valid.
I assume that xsubpp has no way of knowing whether <code>MYDEF123</code> is defined or not. All it can see is that the function is defined twice, and it warns accordingly.
Of course, if <code>MYDEF123</code> is defined, then the warning is definitely justified - and I would certainly be expecting to be told that a function had been defined twice, even (especially ?) when both definitions are identical (as is the case with the demo).

I'm not sure whether @choroba meant for the 2 definitions to be identical.
In practice, you'd expect to find that they're different - but, either way, it makes little difference in the context of this issue.

I think @choroba's demo is an example of a poorly written XS file. It should be rewritten so as to avoid the warning:
```
MODULE = My PACKAGE = My

void
foo(dbh)
   SV *dbh
CODE:
{
#ifdef MYDEF123
   int x;
   ++x;
#else
   int x;
   ++x;
#endif
}
``` 
Cheers,
Rob

**Comment by choroba** [comment]
That's the problem of simplified examples. In the original issue, the definitions were in different files, so ifdefing just the function body wasn't possible.

**Comment by sisyphus** [comment]
> In the original issue, the definitions were in different files, so ifdefing just the function body wasn't possible.

Aaah, yes - I see.
However, I still feel that xsubpp is doing the right thing .
If the warnings are to be avoided, then I think it's the responsibility of the DBD-MariaDB source, not of EU::ParseXS.

**Comment by pali** [comment]
As said this is problem of simplified example. Please look into the linked DBD-MariaDB issue. And do not say that this is DBD-MariaDB issue if you even have not properly investigated whole DBI code which is statically linked into DBD modules. Problem is not in DBD-MariaDB, it is conditionally defining XS function which is not available in older DBI versions. Makefile for DBD-MariaDB is properly adding preprocessor define macro which is evaluated during compilation, but looks like that is not properly respected or evaluated by XS compiler. At the end this warning is not something which can be fixed or workarounded in DBD-MariaDB.

**Comment by sisyphus** [comment]
On Tue, Jul 5, 2022 at 7:44 PM pali ***@***.***> wrote:
  .....

> DBD-MariaDB issue. And do not say that this is DBD-MariaDB issue if you
> even have not properly investigated whole DBI code which is statically
> linked into DBD modules.
>
Sounds like a fair point.
I certainly have not investigated the DBI code, nor do I have any intention
of doing so.
And I probably should have thought of the possibility that it not being an
EU::ParseXS issue doesn't necessarily mean that it's a DBD-MariaDB issue.

Cheers,
Rob

**Comment by pali** [comment]
In my opinion this is just giant layering problem due to how many processing parts are involved during compilation of DBD drivers generally (which applies also for DBD-MariaDB). At upper level are templated XS files and lower level are expanded C files (when C preprocessor finish expansion of all macros). Between these two ends are DBI template expander, XS compilers, C preprocessors, Makefiles, etc... I have no idea how to handle such issue.

**Comment by demerphq** [comment]
@sisyphus - I see where you coming from that xsubpp is doing the right thing, but it feels to me like that would only be true if it is xsubpp's *job* to detect duplicate definitions and I dont think it is. I would have said that was the compilers job. And especially in a case where both definitions are identical, it seems like xsubpp should just say "if this is a problem a different part of the toolchain will deal with it".

I think we should fix xsubpp to pass @jkeenan's tests. 

Although i do wonder about @choroba's example. Is the code involved really doing:
```
#ifdef x
SUBDEF
#endif
SUBDEF
```
or is it doing:
```
#ifdef x
SUBDEF
#else
SUBDEF
#endif
```
if we are warning on the latter then IMO it definitely is a bug.

**Comment by pali** [comment]
Please, really look at the original issue and into DBI and DBD-MariaDB module source code. @choroba's example is really simplified and probably it does not make sense to spend time with academic discussion aboutthat  simplified issue which is not the real issue. DBD drivers use another XS layer from DBI framework and therefore during DBD driver build process are many more include layers. So this is what needs to be inspected by somebody with xs/xsubpp knowledge.

**Comment by sisyphus** [comment]
> I see where you coming from that xsubpp is doing the right thing, but it feels to me like that would only be true if it is xsubpp's job to detect duplicate definitions and I dont think it is.

@demerphq - if it's doing something that it has no business doing (and I now think you're right about that), then that's easily fixed.
We have, beginning at line 356 of ParseXS.pm version 3.45 (perl-5.36.0):
```
    # Check for duplicate function definition
    for my $tmp (@{ $self->{XSStack} }) {
      next unless defined $tmp->{functions}{ $self->{Full_func_name} };
      Warn( $self, "Warning: duplicate function definition '$clean_func_name' detected");
      last;
    }
```
Just comment those lines out, and no more warnings. I've tested this "solution", and it works fine.

But if we want xsubpp to warn only for cases that will cause a compiler error, but have it not warn for cases that won't cause a compiler error, then we need to make it a bit smarter.
I don't have another handy little glib solution for that case ;-)

AFAICS there's no point in xsubpp warning about this if the compiler is only going to throw an error, anyway.
And if the compiler is not going to throw an error, then there's actually nothing to warn about, and the warning is therefore pointless ??
So, why not simply comment out that check in ParseXS.pm ?
Is there's something I'm overlooking ? 

> Although i do wonder about @choroba's example.

I've checked, and it (correctly) doesn't warn for the "latter case".
So xsubpp is smart enough to know that the compiler is not going to process the code in both the \#ifdef and the \#else block.
But, AFAICT, it *never* knows which of the 2 blocks will be processed - even if the XS file has explicitly defined which block  to process. 

Cheers,
Rob

**Comment by demerphq** [comment]
@sisyphus I agree with your analysis pretty much although I want to poke into ParseXS a bit more before i say anything more.  I have been finding it a bit confounding to wrap my head around.

**Comment by demerphq** [comment]
@pali its not that helpful to tell us to just ignore this given that @choroba gave us an example that produces similar output. if you think there is a more representative case to look into then could you try to produce a reduction we can debug? At the lowest level xsubpp reads some files and either does the right thing or not right? So "all the layers" shouldn't matter.  There must exist a single layer that ends up with weird stuff and we can look into whether that weird stuff is being processed helpfully or not.

**Comment by pali** [comment]
I did not wanted to sounds like ignoring issue. Just wanted to point out that due to simplified example, it could be different issue as the original which is happening during building of DBI.

**Comment by iabyn** [comment]
This issue should be fixed in the xsubpp 3.61 just merged into bleadperl by v5.43.5-177-g528de8c84b

**Comment by choroba** [comment]
For the curious, this is the commit: 528de8c84b56b02abee8449ae267e17bf1a6ee8f

**Comment by twata1** [comment]
It was worth reporting. Thank you.

I have confirmed that the issue in question has been resolved on Windows Perl 5.43.6 compared to the results on Windows Perl 5.43.2.

Before (Windows Perl 5.43.2): 
https://www.cpantesters.org/cpan/report/6acb99fe-6c22-1014-a72c-b08793dbee25

After: (Windows Perl 5.43.6):
https://www.cpantesters.org/cpan/report/68029b7e-6c23-1014-a055-e11a8b338db2

It seems the updates haven't caught up, but the following site may be easier to understand as the warning lines are highlighted in color.
https://matrix.perl-magpie.org/dist/DBD-MariaDB

UPDATE:
Before (Windows Perl 5.43.2):
https://matrix.perl-magpie.org/results/044821e6-6c9b-1014-9b3d-cdb49c46273f

After: (Windows Perl 5.43.6):
https://matrix.perl-magpie.org/results/461a6b83-6c9b-1014-bd23-d9df7108b779

(Regarding these test reports, the build itself will likely fail because it is attempting to link 64-bit libraries with a 32-bit compiler.)

Thank you,
